taichi/transforms/constant_fold.cpp

```diff
@@ -11,61 +14,161 @@ class ConstantFold : public BasicStmtVisitor {
   ConstantFold() : BasicStmtVisitor() {
   }
 
-  void visit(UnaryOpStmt *stmt) override {
-    if (stmt->width() == 1 && stmt->op_type == UnaryOpType::cast_value &&
-        stmt->operand->is<ConstStmt>()) {
-      auto input = stmt->operand->as<ConstStmt>()->val[0];
-      auto src_type = stmt->operand->ret_type.data_type;
-      auto dst_type = stmt->ret_type.data_type;
-      TypedConstant new_constant(dst_type);
-      bool success = false;
-      if (src_type == DataType::f32) {
-        auto v = input.val_float32();
-        if (dst_type == DataType::i32) {
-          new_constant.val_i32 = int32(v);
-          success = true;
-        }
-      } else if (src_type == DataType::i32) {
-        auto v = input.val_int32();
-        if (dst_type == DataType::f32) {
-          new_constant.val_f32 = float32(v);
-          success = true;
+  struct JITEvaluatorId
+  {
+    int op;
+    DataType ret, lhs, rhs;
+    bool is_binary;
+
+    explicit operator int() const // make STL map happy
+    {
+      return (int)op | (int)!!is_binary << 7
+        | (int) ret << 8 | (int) lhs << 16 | (int) rhs << 24;
+    }
+
+    UnaryOpType unary_op() const
+    {
+      TI_ASSERT(!is_binary);
+      return (UnaryOpType) op;
+    }
+
+    BinaryOpType binary_op() const
+    {
+      TI_ASSERT(is_binary);
+      return (BinaryOpType) op;
+    }
+  };
+
+  static Kernel *get_jit_evaluator_kernel(JITEvaluatorId const &id)
+  {
+    auto &cache = get_current_program().jit_evaluator_cache;
+    int iid = int(id);
+    auto it = cache.find(iid);
+    if (it != cache.end()) // cached?
+      return it->second.get();
+    static int jic = 0; // X: race?
```

Comment:
Let's make it an `std::atomic<uint64>` member of `Program`.

Reply:
Here I used `cache.size()` instead of `jic`, and I realized that we may need a mutex for `cache`?